hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/distributed/NodeQueueLoadMonitor.java

```diff
@@ -64,40 +69,165 @@
+          final float availableRatio1 =
+              resourceCalculator.minRatio(availableResource1, clusterResource);
+          final float availableRatio2 =
+              resourceCalculator.minRatio(availableResource2, clusterResource);
+
+          // The one with more available resources should be placed first
+          diff = Precision
+              .compareTo(availableRatio2, availableRatio1, Precision.EPSILON);
+        }
+
+        if (diff == 0) {
+          // Compare absolute value if ratios are the same
+          diff = availableResource2.getVirtualCores() - availableResource1.getVirtualCores();
+        }
+
+        if (diff == 0) {
+          diff = Long.compare(availableResource2.getMemorySize(),
+              availableResource1.getMemorySize());
+        }
+        break;
+      case QUEUE_WAIT_TIME:
+      case QUEUE_LENGTH:
+      default:
+        diff = getMetric(o1) - getMetric(o2);
+        break;
+      }
+
+      if (diff == 0) {
+        return (int) (o2.getTimestamp() - o1.getTimestamp());
       }
-      return getMetric(o1) - getMetric(o2);
+      return diff;
+    }
+
+    private void setClusterResource(Resource clusterResource) {
+      this.clusterResource = clusterResource;
+    }
+
+    public ResourceCalculator getResourceCalculator() {
+      return resourceCalculator;
     }
 
     public int getMetric(ClusterNode c) {
-      return (this == QUEUE_LENGTH) ?
-          c.getQueueLength().get() : c.getQueueWaitTime().get();
+      switch (this) {
+      case QUEUE_WAIT_TIME:
+        return c.getQueueWaitTime().get();
+      case QUEUE_LENGTH:
+      case QUEUE_LENGTH_THEN_RESOURCES:
+      default:
+        return c.getQueueLength().get();
+      }
     }
 
     /**
      * Increment the metric by a delta if it is below the threshold.
      * @param c ClusterNode
      * @param incrementSize increment size
+     * @param requested the requested resource
      * @return true if the metric was below threshold and was incremented.
      */
-    public boolean compareAndIncrement(ClusterNode c, int incrementSize) {
-      if(this == QUEUE_LENGTH) {
-        int ret = c.getQueueLength().addAndGet(incrementSize);
-        if (ret <= c.getQueueCapacity()) {
+    public boolean compareAndIncrement(
+        ClusterNode c, int incrementSize, Resource requested) {
+      if (this == QUEUE_LENGTH_THEN_RESOURCES) {
+        // Assignment and getting value is atomic
+        // Can be slightly inaccurate here, don't grab lock for performance
+        final Resource capability = c.getCapability();
+        final Resource currAllocated = c.getAllocatedResource();
+        final Resource currAvailable = Resources.subtractNonNegative(
+            capability, currAllocated);
+        if (resourceCalculator.fitsIn(requested, currAvailable)) {
+          final Resource newAllocated = Resources.add(currAllocated, requested);
+          c.setAllocatedResource(newAllocated);
```

Comment:
with multiple application running in parallel the allocated resource update could be wrong here. Do we take clusterNode level lock ??

Reply:
Yeah, I saw that since the original code also wasn't handling parallelism correctly I thought it might've been fine to not use locks, but ensuring correctness is probably better here.